tracing/src/main/java/com/palantir/tracing/Trace.java

```diff
@@ -28,62 +34,214 @@
 
-    private final Deque<OpenSpan> stack;
-    private final boolean isObservable;
     private final String traceId;
 
-    private Trace(ArrayDeque<OpenSpan> stack, boolean isObservable, String traceId) {
-        checkArgument(!traceId.isEmpty(), "traceId must be non-empty");
-
-        this.stack = stack;
-        this.isObservable = isObservable;
+    private Trace(String traceId) {
+        checkArgument(!Strings.isNullOrEmpty(traceId), "traceId must be non-empty");
         this.traceId = traceId;
     }
 
-    Trace(boolean isObservable, String traceId) {
-        this(new ArrayDeque<>(), isObservable, traceId);
+    /**
+     * Opens a new span for this thread's call trace, labeled with the provided operation and parent span. Only allowed
+     * when the current trace is empty.
+     * If the return value is not used, prefer {@link #fastStartSpan(String, String, SpanType)}}.
+     */
+    @CheckReturnValue
+    final OpenSpan startSpan(String operation, String parentSpanId, SpanType type) {
+        checkState(isEmpty(), "Cannot start a span with explicit parent if the current thread's trace is non-empty");
+        checkArgument(!Strings.isNullOrEmpty(parentSpanId), "parentSpanId must be non-empty");
+        OpenSpan span = OpenSpan.of(operation, Tracers.randomId(), type, Optional.of(parentSpanId));
+        push(span);
+        return span;
     }
 
-    void push(OpenSpan span) {
-        stack.push(span);
+    /**
+     * Opens a new span for this thread's call trace, labeled with the provided operation.
+     * If the return value is not used, prefer {@link #fastStartSpan(String, SpanType)}}.
+     */
+    @CheckReturnValue
+    final OpenSpan startSpan(String operation, SpanType type) {
+        Optional<OpenSpan> prevState = top();
+        final OpenSpan span;
+        // Avoid lambda allocation in hot paths
+        if (prevState.isPresent()) {
+            span = OpenSpan.of(operation, Tracers.randomId(), type, Optional.of(prevState.get().getSpanId()));
+        } else {
+            span = OpenSpan.of(operation, Tracers.randomId(), type, Optional.empty());
+        }
+
+        push(span);
+        return span;
     }
 
-    Optional<OpenSpan> top() {
-        return stack.isEmpty() ? Optional.empty() : Optional.of(stack.peekFirst());
-    }
+    /**
+     * Like {@link #startSpan(String, String, SpanType)}, but does not return an {@link OpenSpan}.
+     */
+    abstract void fastStartSpan(String operation, String parentSpanId, SpanType type);
 
-    Optional<OpenSpan> pop() {
-        return stack.isEmpty() ? Optional.empty() : Optional.of(stack.pop());
-    }
+    /**
+     * Like {@link #startSpan(String, SpanType)}, but does not return an {@link OpenSpan}.
+     */
+    abstract void fastStartSpan(String operation, SpanType type);
 
-    boolean isEmpty() {
-        return stack.isEmpty();
-    }
+    abstract void push(OpenSpan span);
+
+    abstract Optional<OpenSpan> top();
+
+    abstract Optional<OpenSpan> pop();
```

Comment:
I find the contract of these two methods a bit confusing... normally an Optional.empty means an element was not present, but in this case the UnsampledTrace _always_ returns Optional.empty?

Reply:
That is correct, this PR introduces a subtle change in behavior where completing an unsampled span (even using the slow complete span method) no longer returns a present value, because we know it is not recorded (in fact, sampling an unsampled span would be a bug).

Our tracing API is missing clean isolation between producers and consumers -- if I am creating spans, I don't need to know whether or not they are sampled, or what the span IDs/timing looks like, just how to start and stop them. Conceptually there's no reason to consume these values because that functionality is a concern of the `SpanObserver`.

The contract is defined by https://github.com/palantir/tracing-java/blob/42aafc8203a96be53ee52794f058d19d54db24fa/tracing/src/main/java/com/palantir/tracing/Tracer.java#L182-L188

There's a lot going on in that comment: "returns the current span (if it exists)". If we're not sampled, we don't really have a current span because we didn't record one. We have to track how many we would have created in the sampled case, but it does not violate the contract to return an empty value. This is reinforced by the next statement: "notifies all subscribers about the completed span", because subscribers are only notified if spans are sampled.

Reply:
Hey, this is a very surprising PR. It is very much a break in behavior, and excusing it on a possible implied reading of the javadoc is not reasonable. This was done as a minor version bump (although given our current tooling a major version bump wouldn't have made a huge difference).

> Conceptually there's no reason to consume these values because that functionality is a concern of the SpanObserver.

If that were true we would only ever be using `fastCompleteSpan()`. As that isn't true, this is not an implementation detail but an API contract.